lib/charms/haproxy/v1/haproxy_route.py

```
@@ -170,6 +171,9 @@ def value_contains_invalid_characters(value: Optional[str]) -> Optional[str]:
     return value
 
 
+VALIDSTR = Annotated[str, BeforeValidator(value_contains_invalid_characters)]
```

In terms of security it's better to explicitely allow some characters. We had a similar discussion for the configurator, you can probably reuse some of the validators here http://github.com/canonical/ingress-configurator-operator/blob/main/src/validators.py#L14

We should also probably have specific validators for the different kind of strings we may have (I'll drop comments below).

---

The validator is not related to the relation data attributes themselves and is mainly to avoid reserved characters in the haproxy config, so that the requirer can't set something like
```
service: "backend\nbackend injected\n\tserver <malicious IP address>"
```
for example. Other reserved character are also part of the banlist as they will break the haproxy config.

---

My doubt is: are we sure we are exhaustive in the banlist?

---

I got the banlist from the list of characters that need to be quoted in the haproxy documentation : https://www.haproxy.com/documentation/haproxy-configuration-manual/latest/#2.2

---

so basically `<space> , ' " # \ \n \r \t` and I skipped ascii hex code for now as it's too complicated to implement.

---

Now that I think about it, as this is specific to haproxy, we can also move the validation to the haproxy charm, and leave room for more meaningful validation in the lib. WDYT?

---

I guess it depends on where we want to raise the potential errors. If we keep it in the library, could it raise error on the requirer side? I feel that would be better in terms of user experience that having the validation (silently) failing on the provider side.

---

Yes having them here will make it so that `DataValidationError` is raised when calling `provide_haproxy_route_requirements` with invalid haproxy characters

---

in that case I'd keep it like this.